### DISCUSSION TITLE
Is it possible to link Jamulus to a webpage

### DISCUSSION THREAD

**Discussion by mksteve** [jamulussoftware/jamulus#1225] [Ideas]
My guitar group are going to struggle to set up and connect with Jamulus.  If I could use webaudio/websockets/webrtc to send to Jamulus, and the machine had low enough latency in its browser and audio hardware, this would be possible.

Having a browser based client, would simplify the use case for a lot of users.

A quick test on my laptop, had websocket text transmission running at about 20ms send-return.  This is not ideal, but just about within spec.

Has this been investigated?
Is this possible?

**Comment by ann0see** [comment]
I'm not sure, but of course you could write a web client (but I doubt latency would be acceptable) honestly. 

> My guitar group are going to struggle to set up and connect with Jamulus. 

I'd focus on this issue. Do they have problems with ASIO4ALL setup? If you can avoid it, avoid it (by buying an USB interface/switching to Jamulus OS which you can run from a USB stick).

**Comment by hoffie** [comment]
Having a browser-based client is certainly doable in principle (although with some effort!) and it would require backend support (e.g. websocket support in Jamulus or a proxy between the browser and a Jamulus server).

Using the MediaRecorder API will probably not yield low enough latency (one would have to record in chunks and the fastest mode I got working was about ~100ms, depending on the browser; this does not cover network delay yet).
I don't know about the latency with WebRTC, maybe it could work, but this would also mean that you would need a WebRTC bridge in the backend.
Would be interesting to see when someone comes up with a PoC :)

My gut feeling is that it could be hard, latency-wise, but I'd welcome to be proven wrong :)

**Reply by Lwnzhn** [reply]
For a very fast implementation of webRTC please check out digital-stage.org -web version. Open Source on Github. Spoiler: WebRTC is not good for <50ms end to end latency.
